tests/engine/training/test_graph_trainer.py

```diff
@@ -0,0 +1,236 @@
+from pathlib import Path
+from typing import Callable, Dict, Optional, Text, Type
+from unittest.mock import Mock
+
+from _pytest.monkeypatch import MonkeyPatch
+from _pytest.tmpdir import TempPathFactory
+import pytest
+
+from rasa.engine.caching import TrainingCache
+from rasa.engine.graph import GraphComponent, GraphSchema, SchemaNode
+from rasa.engine.runner.dask import DaskGraphRunner
+from rasa.engine.storage.local_model_storage import LocalModelStorage
+from rasa.engine.storage.resource import Resource
+from rasa.engine.storage.storage import ModelStorage
+from rasa.engine.training.graph_trainer import GraphTrainer
+from tests.engine.graph_components_test_classes import (
+    AddInputs,
+    AssertComponent,
+    FileReader,
+    PersistableTestComponent,
+)
+
+
+def test_graph_trainer(
+    default_model_storage: ModelStorage,
+    temp_cache: TrainingCache,
+    tmp_path: Path,
+    domain_path: Path,
+):
+    graph_trainer = GraphTrainer(
+        model_storage=default_model_storage,
+        cache=temp_cache,
+        graph_runner_class=DaskGraphRunner,
+    )
+
+    test_value = "test_value"
+
+    train_schema = GraphSchema(
+        {
+            "train": SchemaNode(
+                needs={},
+                uses=PersistableTestComponent,
+                fn="train",
+                constructor_name="create",
+                config={"test_value": test_value,},
+                is_target=True,
+            ),
+            "load": SchemaNode(
+                needs={"resource": "train"},
+                uses=PersistableTestComponent,
+                fn="run_inference",
+                constructor_name="load",
+                config={},
+            ),
+        }
+    )
+    predict_schema = GraphSchema(
+        {
+            "load": SchemaNode(
+                needs={},
+                uses=PersistableTestComponent,
+                fn="run_inference",
+                constructor_name="load",
+                config={},
+                is_target=True,
+                resource=Resource("train"),
+            ),
+        }
+    )
+
+    output_filename = tmp_path / "model.tar.gz"
+    predict_graph_runner = graph_trainer.train(
+        train_schema=train_schema,
+        predict_schema=predict_schema,
+        domain_path=domain_path,
+        output_filename=output_filename,
+    )
+    assert isinstance(predict_graph_runner, DaskGraphRunner)
+    assert output_filename.is_file()
+    assert predict_graph_runner.run() == {"load": test_value}
+
+
+def test_graph_trainer_fingerprints_and_caches(
```

Comment:
how about a test 
* where no component returns a cacheable result
* a file content changes and hence makes the training required